### PR TITLE
Report nullability mismatches for delegate creation expressions

### LINKED ISSUE #32499: Roslyn doesn't verify nullable reference types in a signature of a delegate constructor argument
**Version Used**: 

Microsoft Visual Studio Enterprise 2019 Preview
Version 16.0.0 Preview 1.1
VisualStudio.16.Preview/16.0.0-pre.1.1+28408.50
Microsoft .NET Framework
Version 4.7.03056

**Steps to Reproduce**:

Compile and run the following code
```cs
#nullable enable
delegate void MyDelegate(C? c);
class C
{
  public static void Main()
  {
    var action = new MyDelegate(M);
    action(null);
  }

  static void M(C c)
  {
    c.ToString();
  }
}
```

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOyQExAagB8BiTdAG3ImHLgAI5NraBYAKBzloHMJ46AAgBY6AWQCeAES5xe8ABQBhAPx0AxgEoA3OwEAmOovYBvdnUEBmQQEYAbIJGiIAS0zyNZuqbbnzANwgEOgg1GGcAe0w6AF46TDgAdzEpGTk4eVFtT3MQsMj5MkosnzoAX3ZPATsHMSV1DxLvX3UAOgAVcIBlGARXbncdEvK2UqA=

**Expected Behavior**:
`warning CS8622: Nullability of reference types in type of parameter 'c' of 'void C.M(C c)' doesn't match the target delegate 'MyDelegate'.`
Just as it would have been reported for `MyDelegate action = M;`

**Actual Behavior**:
No warnings. The program crashes at runtime with a `NullReferenceException`

### LINKED ISSUE #32698: Need nullability analysis of delegate creation expressions
```C#
        [Fact]
        public void SuppressNullableWarning_DelegateCreation()
        {
            var source =
@"class C
{
    static void Main()
    {
        _ = new System.Func<string, string>((string? x) => { return null; }!);
        _ = new System.Func<string?, string?>((string x) => { return string.Empty; }!);
        _ = new System.Func<string, string>(M1!);
        _ = new System.Func<string?, string?>(M2!);

        // without suppression
        _ = new System.Func<string, string>((string? x) => { return null; }); // 1
        _ = new System.Func<string?, string?>((string x) => { return string.Empty; }); // 2
        _ = new System.Func<string, string>(M1); // 3
        _ = new System.Func<string?, string?>(M2); // 4
    }
    static string? M1(string? x) => throw null;
    static string M2(string x) => throw null;
}";

            // missing warnings

            // cover cases in BindDelegateCreationExpression
            var comp = CreateCompilation(source, options: WithNonNullTypesTrue(TestOptions.DebugExe));
            comp.VerifyDiagnostics(
                );
            CompileAndVerify(comp);

            // Need to verify the semantic model
            // Tracked by https://github.com/dotnet/roslyn/issues/32661
        }
```

### PR DESCRIPTION
**Pull request by cston**
Fixes #32499
Fixes #32698

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -4194,7 +4194,6 @@ private bool HasTopLevelNullabilityConversion(TypeWithAnnotations source, TypeWi
                     {
                         var lambda = (BoundLambda)operandOpt;
                         var delegateType = targetType.GetDelegateType();
-                        var unboundLambda = lambda.UnboundLambda;
                         var variableState = GetVariableState(stateForLambda);
                         Analyze(compilation,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >Analyze(compilation, [](start = 24, length = 20)

Could this also be just a VisitLambda call?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -4599,22 +4598,43 @@ private TypeSymbol MakeNullableOf(TypeWithAnnotations underlying)
...
-            else
-            {
-                VisitRvalue(node.Argument);
+                case BoundMethodGroup group:
+                    {
+                        VisitRvalue(group.ReceiverOpt);
+                        SetAnalyzedNullability(group, default);
+                        // https://github.com/dotnet/roslyn/issues/33637: Should update method based on inferred receiver type.
+                        var method = node.MethodOpt;
+                        if (!(method is null) && !group.IsSuppressed)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >!(method is null) [](start = 28, length = 17)

```method is object```?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -2181,12 +2181,26 @@ static void Main()
     static string M2(string x) => throw null!;
 }";
 
-            // missing warnings
-            // Tracked by https://github.com/dotnet/roslyn/issues/32698
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >// Tracked by https://github.com/dotnet/roslyn/issues/32698 [](start = 12, length = 59)

Can  [https://github.com/dotnet/roslyn/issues/32698](https://github.com/dotnet/roslyn/issues/32698) also be resolved?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -2181,12 +2181,26 @@ static void Main()
...
-
             var comp = CreateCompilation(source, options: WithNonNullTypesTrue(TestOptions.DebugExe));
             comp.VerifyDiagnostics(
-                );
+                // (5,69): warning CS8603: Possible null reference return.
+                //         _ = new System.Func<string, string>((string? x) => { return null; }!);
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(5, 69),
+                // (11,57): warning CS8622: Nullability of reference types in type of parameter 'x' of 'lambda expression' doesn't match the target delegate 'Func<string, string>'.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >// (11,57): warning CS8622: Nullability of reference types in type of parameter 'x' of 'lambda expression' doesn't match the target delegate 'Func<string, string>'. [](start = 16, length = 164)

It feels like this warning is unexpected.

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticAnalyzerTests.cs [hunk 5]
```diff
@@ -77,21 +77,20 @@ static void Main()
...
-// (9,42): warning CS1720: Expression will always cause a System.NullReferenceException because the default value of 'Program.C' is null
-//         var o2 = new D(delegate{ var s = default(C).ToString();});
-Diagnostic(ErrorCode.WRN_DotOnDefault, "default(C).ToString").WithArguments("Program.C")
-                );
+                // (8,40): warning CS1720: Expression will always cause a System.NullReferenceException because the default value of 'Program.C' is null
+                //         var o1 = (D)(delegate{ var s = default(C).ToString();});
+                Diagnostic(ErrorCode.WRN_DotOnDefault, "default(C).ToString").WithArguments("Program.C"),
+                // (9,42): warning CS1720: Expression will always cause a System.NullReferenceException because the default value of 'Program.C' is null
+                //         var o2 = new D(delegate{ var s = default(C).ToString();});
+                Diagnostic(ErrorCode.WRN_DotOnDefault, "default(C).ToString").WithArguments("Program.C"));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SemanticAnalyzerTests.cs]: >")); [](start = 102, length = 4)

Consider preserving original formatting (the ```);``` on a separate line).

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticAnalyzerTests.cs [hunk 6]
```diff
@@ -77,21 +77,20 @@ static void Main()
...
             CreateCompilation(source, options: TestOptions.ReleaseDll.WithNullableContextOptions(NullableContextOptions.SafeOnlyWarnings)).VerifyDiagnostics(
                 // (8,40): warning CS8602: Dereference of a possibly null reference.
                 //         var o1 = (D)(delegate{ var s = default(C).ToString();});
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "default(C)").WithLocation(8, 40)
-                // Missing warning due to https://github.com/dotnet/roslyn/issues/32698
-                );
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "default(C)").WithLocation(8, 40),
+                // (9,42): warning CS8602: Dereference of a possibly null reference.
+                //         var o2 = new D(delegate{ var s = default(C).ToString();});
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "default(C)").WithLocation(9, 42));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SemanticAnalyzerTests.cs]: >)); [](start = 96, length = 3)

Consider preserving original formatting (the ); on a separate line).

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 4 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: The behavior is consistent with the implicit conversion, and with the similar conversion behavior when using base and derived types:
```
System.Func<string, string> f =
    (string? x) => { return null; }; // warning
_ = new System.Func<string, string>(
    (object x) => { return ""; }); // warning
```
We can relax the requirement that parameter nullability matches but I'd prefer to do so in a separate PR.

---
In reply to: [281751757](https://github.com/dotnet/roslyn/pull/35547#discussion_r281751757) [](ancestors = 281751757)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Sure, although `VisitLambda` will need an optional `initialState` parameter.

---
In reply to: [281745456](https://github.com/dotnet/roslyn/pull/35547#discussion_r281745456) [](ancestors = 281745456)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 4 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Logged [https://github.com/dotnet/roslyn/issues/35564](https://github.com/dotnet/roslyn/issues/35564)

---
In reply to: [281757577](https://github.com/dotnet/roslyn/pull/35547#discussion_r281757577) [](ancestors = 281757577,281751757)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Seems equivalent.

---
In reply to: [281745786](https://github.com/dotnet/roslyn/pull/35547#discussion_r281745786) [](ancestors = 281745786)

**Pull request merged**